eth_defi/token_tax.py

```diff
@@ -0,0 +1,134 @@
+from email.quoprimime import quote
+from eth_account.signers.local import LocalAccount
+from eth_typing import HexAddress, HexStr
+from eth_defi.uniswap_v3.constants import FOREVER_DEADLINE
+from web3 import Web3
+
+from eth_defi.uniswap_v2.deployment import UniswapV2Deployment
+from eth_defi.abi import get_deployed_contract
+from eth_defi.token import fetch_erc20_details
+
+from dataclasses import dataclass
+
+
+@dataclass
+class TokenTaxInfo:
+    """Different token taxes we figured out."""
+
+    #: Token in the question
+    base_token: HexAddress
+
+    #: Which token we traded against it
+    quote_token: HexAddress
+
+    #: How much % we lost of the token on buy
+    buy_tax: float
+
+    #: How much % we lose the token when we transfer between addresses
+    transfer_tax:float
+
+    #: How much % we lose the token when we sold it
+    sell_tax: float
+
+
+def estimate_token_taxes(
+        uniswap: UniswapV2Deployment,
+        base_token: HexAddress,
+        quote_token: HexAddress,
+        buy_account: HexAddress,
+        sell_account: HexAddress
+    ) -> TokenTaxInfo:
+    """Estimates different token taxes for a token by running Ganache simulations for it.
+
+    :param uniswap:
+        Uniswap deployment on a Ganache mainnet fork.
+        Set up prior calling this function.
+        See `ganache.py` and `test_ganache.py` for more details.
+
+    :param base_token:
+        The token of which tax properties we are figuring out.
+
+    :param quote_token:
+        Address of the quote token used for the trading pair. E.g. `BUDS`, `WBNB`
+        Based on this information we can derive Uniswap trading pair address.
+
+    :param buy_account:
+        The account that does initial buy to measure the buy tax.
+        This account must be loaded with gas money (ETH/BNB) and `quote_token`
+        for a purchase.
+
+    :param sell_account:
+        The account that receives the token transfer and does the sell to measure the sell tax.
+        This account must be loaded with gas money for the sell.
+
+    :return:
+        ToxTaxInfo tells us what we figure out about taxes.
+        This can be later recorded to a database.
+    """
+
+    web3: Web3 = uniswap.web3
+    router = uniswap.router
+
+    quote_token_details = fetch_erc20_details(web3, quote_token)
+    quote_token = quote_token_details.contract
+
+    base_token_details = fetch_erc20_details(web3, base_token)
+    base_token = base_token_details.contract
+
+    # approve router to spend tokens
+    quote_token.functions.approve(router.address, quote_token_details.convert_to_raw(99999)).transact({"from": buy_account})
```

Comment:
Don't assume hard coded 9999 parameters here. Instead, move whatever test buy amount me to be a default argument of the function, so the user can adjust it easily.